internal/service/dynamodb/global_secondary_index.go

```diff
@@ -0,0 +1,710 @@
+	awstypes "github.com/aws/aws-sdk-go-v2/service/dynamodb/types"
+	"github.com/dlclark/regexp2"
+	"github.com/hashicorp/aws-sdk-go-base/v2/tfawserr"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/customdiff"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/retry"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-provider-aws/internal/conns"
+	"github.com/hashicorp/terraform-provider-aws/internal/create"
+	"github.com/hashicorp/terraform-provider-aws/internal/enum"
+	"github.com/hashicorp/terraform-provider-aws/internal/errs"
+	"github.com/hashicorp/terraform-provider-aws/internal/flex"
+	"github.com/hashicorp/terraform-provider-aws/internal/slices"
+	"github.com/hashicorp/terraform-provider-aws/names"
+)
+
+const (
+	resNameGlobalSecondaryIndex = "global_secondary_index"
+)
+
+// @SDKResource("aws_dynamodb_global_secondary_index", name="Global Secondary Index")
+// @Testing(existsType="github.com/aws/aws-sdk-go-v2/service/dynamodb/types;types.TableDescription")
+func resourceGlobalSecondaryIndex() *schema.Resource {
+	//lintignore:R011
+	return &schema.Resource{
+		CreateWithoutTimeout: resourceGlobalSecondaryIndexCreate,
+		ReadWithoutTimeout:   resourceGlobalSecondaryIndexRead,
+		UpdateWithoutTimeout: resourceGlobalSecondaryIndexUpdate,
+		DeleteWithoutTimeout: resourceGlobalSecondaryIndexDelete,
+
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(createTableTimeout),
+			Delete: schema.DefaultTimeout(deleteTableTimeout),
+			Update: schema.DefaultTimeout(updateTableTimeoutTotal),
+		},
+
+		SchemaVersion: 1,
+
+		Schema: map[string]*schema.Schema{
+			names.AttrIndexARN: {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"table": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"hash_key": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"hash_key_type": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+			"range_key": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"range_key_type": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+			names.AttrName: {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"non_key_attributes": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				ForceNew: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			"on_demand_throughput": onDemandThroughputSchema(),
+			"projection_type": {
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				ValidateDiagFunc: enum.Validate[awstypes.ProjectionType](),
+			},
```

Comment:
To match the underlying API, `projection_type` and `non_key_attributes` should be moved into a nested block `projection`